specs/capri/Multi-attach.md

```diff
@@ -0,0 +1,151 @@
+
+## Summary
+
+This design is proposed to enable multiAttach of volumes.
+
+## Motivation
+
+Currently we refuse to attach a volume that is already attached (in-use). There are situations where a cloud user may have deployed a clustered file-system on a Cinder volume and they would like the ability to attach a volume to multiple hosts/servers at the same time.
+
+### Goals
+
+This spec proposes a set of changes to enable and control the use of multiAttach volumes from the OpenSDS side.
+
+
+## Design Details
+
+1. The backend devices that support multi-attach ability need to report this in their capabilities in yaml configuration file. And pools belong to this backend should inherit the capability, because multi-attach volume creation request will select multi-attach enabled pool from pools whch come from different backend that may not support this ability. 
+2. Attach a volume as one would normally do (Read Write access) to a host and then attach it to another host in Read Only mode or Read Write Mode. In order to implement this, add an item AttachMode to the volume attachment which has two values: RO (Read Only) and RW (Read Write), default is RW. 
+3. Add multiAttach bool to the volume model to enables/disables the ability of multi-attach. 
+4. Add 'multiAttach': '<is> True' to the profile and do some corresponding changes to the controller to choose backend if user require a multi-attach volume.
+
+### Data model impact
+
+1. Add MultiAttach field in dock spec and dock will load configuration file that has the key value pair multiAttach:true/false and then fill the field.
+```
+type DockSpec struct {
+	*BaseModel
+	Type string `json:"type,omitempty"`
+	Name string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	Status string `json:"status,omitempty"`
+	StorageType string `json:"storageType,omitempty"`
+	Endpoint string `json:"endpoint,omitempty"`
+	NodeId string `json:"nodeId,omitempty"`
+	DriverName string `json:"driverName,omitempty"`
+	MultiAttach bool `json:"multiAttach,omitempty"`
+	Metadata map[string]string `json:"metadata,omitempty"`	
+}
+```
+
+2. Add MultiAttach field in pool spec
+
+```
+type StoragePoolSpec struct {
+	*BaseModel
+	Name string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	Status string `json:"status,omitempty"`
+	DockId string `json:"dockId,omitempty"`
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+	TotalCapacity int64 `json:"totalCapacity,omitempty"`
+	FreeCapacity int64 `json:"freeCapacity,omitempty"`
+	StorageType string `json:"storageType,omitempty"`
+	Extras StoragePoolExtraSpec `json:"extras,omitempty"`
+	ReplicationType string `json:"replicationType,omitempty"`
+	ReplicationDriverName string `json:"replicationDriverName,omitempty"`
+
+	// MultiAttach
+	// If true, this volume can attach to more than one instance. Default will be multiAttach:False
+	MultiAttach bool `json:"multiAttach,omitempty"`
```

Comment:
What's the value of `MultiAttach` if pool's storage type is file?  is it better remove it because `MultiAttach` is always `True`?

Reply:
It cannot be removed and cannot always True.

Reply:
Let's then find a way if pool's storage type is file.

Reply:
If the type of the pool is pool, then analyze